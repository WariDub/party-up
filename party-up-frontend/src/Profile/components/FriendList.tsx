import { Box, Button, Card, CardContent, Grid, GridList, GridListTile } from '@material-ui/core';
import React from 'react';
import User from '../models/user.model';
import { UsersService } from '../users.service';

export interface FriendListProps {
  user: User | null;
  friends: User[];
}

export interface FriendListState {}

class FriendList extends React.Component<FriendListProps, FriendListState> {
  usersService = new UsersService();

  constructor(props: FriendListProps) {
    super(props);
  }

  render(): JSX.Element {
    return (
      <Box m={1}>
        <Grid container justify="center">
          <Grid item xs={8}>
            {this.renderFriends()}
          </Grid>
        </Grid>
      </Box>
    );
  }

  renderFriends(): JSX.Element {
    const { friends } = this.props;

    return (
      <GridList spacing={8} cellHeight="auto" cols={0}>
        {friends.map((friend) => (
          <GridListTile key={friend._id}>{this.renderFriend(friend)}</GridListTile>
        ))}
      </GridList>
    );
  }

  renderFriend(friend: User): JSX.Element {
    const { user } = this.props;

    return (
      <Card>
        <CardContent>
          <h3>{friend.displayName}</h3>
          <h5>{`@${friend.username}`}</h5>
          <Button
            variant="contained"
            color="primary"
            onClick={this.handleButtonOnClickToggleFriend(friend)}
          >
            {user?.friends.includes(friend._id) ? 'Unfriend' : 'Add Friend'}
          </Button>
        </CardContent>
      </Card>
    );
  }

  handleButtonOnClickToggleFriend = (param: User) => (event: any): void => {
    console.log(event);

    const { user } = this.props;
    if (!user) {
      return;
    }

    const addFriend = !user.friends.includes(param._id);

    if (addFriend) {
      user.friends.push(param._id);
      event.target.innerHTML = 'Unfriend';
      this.usersService.addFriend(user.username, param._id);
    } else {
      user.friends = user.friends.filter((friend) => friend != param._id);
      event.target.innerHTML = 'Add Friend';
      this.usersService.deleteFriend(user.username, param._id);
    }
  };
}

export default FriendList;
