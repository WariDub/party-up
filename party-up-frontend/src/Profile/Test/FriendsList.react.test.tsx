import  FriendList from '../components/FriendList';
import React from "react";
import { act } from "react-dom/test-utils";
import { shallow, mount } from 'enzyme';
import { withRouter } from 'react-router';
import User from '../models/user.model';
import Gender from '../enums/gender.enum';
import Genre from '../enums/genre.enum';
import Role from '../enums/role.enum';
import toJson, { shallowToJson } from 'enzyme-to-json';
import renderer from 'react-test-renderer';


describe('FriendList', () => {
    
    test('It should mount', () => {
        let friend: User = {
            _id: '1',
            username: 'TestUsername',
            displayName: 'TestDisplayName',
            email: 'TestEmail@TestEmail.com',
            avatarUrl: 'TestAvatarURL.com',
            friends: ['Friend1', 'Friend2'],
            age: 23,
            gender: Gender.MALE,
            favoriteGenre: Genre.ACTION,
            favoriteRole: Role.DPS,
            createdAt: 5/23/2021
        }

        let component = shallow(<FriendList friends= { [ friend ]} />);
        expect(component.length).toBe(1);
    });

    test('It should render the FriendsList, with all of the correct properties', () => {
        let friend: User = {
            _id: '1',
            username: 'TestUsername',
            displayName: 'TestDisplayName',
            email: 'TestEmail@TestEmail.com',
            avatarUrl: 'TestAvatarURL.com',
            friends: ['Friend1', 'Friend2'],
            age: 23,
            gender: Gender.MALE,
            favoriteGenre: Genre.ACTION,
            favoriteRole: Role.DPS,
            createdAt: 5/23/2021
        }

        let component = renderer.create(<FriendList friends = { [friend] }/>).toJSON();

        expect(component).toMatchSnapshot();
    });
});