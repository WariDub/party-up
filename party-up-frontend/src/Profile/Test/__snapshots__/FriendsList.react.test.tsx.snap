// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FriendList It should render the FriendsList, with all of the correct properties 1`] = `
<div
  className="MuiBox-root MuiBox-root-1"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-center"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
    >
      <ul
        className="MuiGridList-root"
        style={
          Object {
            "margin": -4,
          }
        }
      >
        <li
          className="MuiGridListTile-root"
          style={
            Object {
              "height": "auto",
              "padding": 4,
              "width": "Infinity%",
            }
          }
        >
          <div
            className="MuiGridListTile-tile"
          >
            <div
              className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
            >
              <div
                className="MuiCardContent-root"
              >
                <h3>
                  TestDisplayName
                </h3>
                <h5>
                  @TestUsername
                </h5>
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
